### PR TITLE
Refactor payment gateway redirects handling

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Follows on from #8568 

The `CheckoutController` was holding a lot of responsibilities for knowing which kinds of payment gateways are available and how to initiate the process of redirecting to the external payment page (if needed). In a single method it had the responsibility for knowing:
- all the payment method types that exist
- which payment methods require external processing and which don't
- what code should be called for each individual payment method to initiate the process of redirecting to the external site

This was being hidden somewhat by the way the logic was tucked away in services. It shouldn't be the `CheckoutController`'s job to know all this stuff!

I've made some interface changes here so that `PaymentMethod` objects now know whether or not they require an external payment process, and know how that process should be started and can return the required URL. The checkout can now *ask* any given payment method if it requires external payment processing or not, and *tell* it to start the process and return the relevant external URL (if needed).

Soundtrack: [Man with the Movie Camera](https://www.youtube.com/watch?v=Oo-GoIX_8BE) :musical_note: 

#### What should we test?

Being redirected to Paypal or Stripe at checkout should work as normal.

#### Release notes

Refactored checkout handling of payment gateway redirects

Changelog Category: Technical changes

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Yay!

File: app/controllers/checkout_controller.rb [hunk 1]
```diff
@@ -190,11 +190,14 @@ def checkout_workflow(shipping_method_id)
...
 
   def redirect_to_payment_gateway
-    redirect_path = Checkout::PaypalRedirect.new(params).path
-    redirect_path = Checkout::StripeRedirect.new(params, @order).path if redirect_path.blank?
-    return if redirect_path.blank?
+    payment_method_id = params.dig(:order, :payments_attributes, 0, :payment_method_id)
+    payment_method = Spree::PaymentMethod.find(payment_method_id)
 
-    render json: { path: redirect_path }, status: :ok
+    redirect_url = payment_method&.external_payment_url(order: @order)
```
[the remark this answers is not in the document]
Comment (Matt-Yorkley) [reply app/controllers/checkout_controller.rb:197]: Ah yeah. I'll tidy this up a bit :ok_hand:

**Comment by mkllnk** [comment]
Controller spec fail. :warning:

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

Tested green-path checkouts, with:
- Stripe SCA
- Paypal

Also checked out somewhat edge cases:
- order cycle closure (when in the `/checkout` page)
- stock restrictions (when in the `/checkout` page)

All good, all as before. Merging :+1:

**Pull request merged**